src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/BuildAssetsUtils.cs

```diff
@@ -22,428 +22,698 @@ namespace NuGet.Commands
+        public static void WriteFiles(IEnumerable<MSBuildOutputFile> files, ILogger log)
+        {
+            foreach (var file in files)
             {
-                var projFileName = Path.GetFileName(request.Project.RestoreMetadata.ProjectPath);
+                if (file.Content == null)
+                {
+                    // Remove the file if the XML is null
+                    FileUtility.Delete(file.Path);
+                }
+                else
+                {
+                    log.LogMinimal(string.Format(CultureInfo.CurrentCulture, Strings.Log_GeneratingMsBuildFile, file.Path));
 
-                targetsPath = Path.Combine(request.RestoreOutputPath, $"{projFileName}.nuget.g.targets");
-                propsPath = Path.Combine(request.RestoreOutputPath, $"{projFileName}.nuget.g.props");
+                    // Create the directory if it doesn't exist
+                    Directory.CreateDirectory(Path.GetDirectoryName(file.Path));
+
+                    // Write out XML file
+                    WriteXML(file.Path, file.Content);
+                }
             }
+        }
+
+        /// <summary>
+        /// Create MSBuild targets and props files.
+        /// Null will be returned for files that should be removed.
+        /// </summary>
+        public static List<MSBuildOutputFile> GenerateMultiTargetFailureFiles(
+            string targetsPath,
+            string propsPath,
+            RestoreOutputType restoreType,
+            ILogger log,
+            CancellationToken token)
+        {
+            XDocument targetsXML = null;
+            XDocument propsXML = null;
 
-            // Non-Msbuild projects should skip targets and treat it as success
-            if (!context.IsMsBuildBased && !ForceWriteTargets())
+            // Create an error file for MSBuild to stop the build.
+            targetsXML = GenerateMultiTargetFrameworkWarning();
+
+            if (restoreType == RestoreOutputType.NETCore)
             {
-                return new MSBuildRestoreResult(targetsPath, propsPath, success: true);
+                propsXML = GenerateEmptyImportsFile();
             }
 
-            // Invalid msbuild projects should write out an msbuild error target
-            if (!targetGraphs.Any())
+            var files = new List<MSBuildOutputFile>()
             {
-                return new MSBuildRestoreResult(targetsPath, propsPath, success: false);
-            }
+                new MSBuildOutputFile(propsPath, propsXML),
+                new MSBuildOutputFile(targetsPath, targetsXML),
+            };
 
-            // Framework -> (targets, props)
-            var buildAssetsByFramework = new Dictionary<NuGetFramework, TargetsAndProps>();
+            return files;
+        }
 
-            // Get assets for each framework
-            foreach (var projectFramework in project.TargetFrameworks.Select(f => f.FrameworkName))
+        public static string ReplacePathsWithMacros(string path)
```

Comment:
Method's name and logic are quite specific to the MSBuild restore scenario. I'd keep it private.